db/migrate/20190617191109_move_awx_credentials_to_authentications.rb

```diff
@@ -0,0 +1,327 @@
+class MoveAwxCredentialsToAuthentications < ActiveRecord::Migration[5.0]
+  class Authentication < ActiveRecord::Base
+    include ActiveRecord::IdRegions
+    self.inheritance_column = :_type_disabled
+    serialize :options
+  end
+
+  # This class is port of the code found in both the awx code base, and the
+  # python cryptography library:
+  #
+  #   https://github.com/pyca/cryptography/blob/ed5a7717/src/cryptography/fernet.py
+  #   https://github.com/ansible/awx/blob/4788f0814/awx/main/utils/encryption.py
+  #
+  # The only real portions from this that are from the `awx` side are for the
+  # inheriting `Fernet256` from the pythons cryptography lib's `Fernet` class:
+  #
+  #   https://github.com/ansible/awx/blob/4788f081/awx/main/utils/encryption.py#L20-L23
+  #
+  # The rest pretty much emulates code from cryptography.py
+  class Fernet256
+    attr_reader :hmac, :cipher, :signing_key, :encryption_key
+
+    InvalidToken = Class.new(ArgumentError)
+
+    # This initialize method is mostly pulled from from the ansible override of
+    # the base Fernet library.
+    #
+    #   https://github.com/ansible/awx/blob/4788f0814/awx/main/utils/encryption.py#L24-L35
+    #
+    # The backend is always going to be `OpenSSL` for us, so that is
+    # irrelevant, but everything else is mostly as it is on the python
+    # equivalent.
+    #
+    # The only diverging change was that we initialize the @cipher and @hmac (for
+    # verification) here instead of in the methods that use them (found in the
+    # `cryptography` base class).
+    #
+    def initialize(key)
+      decoded_key = Base64.urlsafe_decode64(key)
+      if decoded_key.size != 64
+        raise ArgumentError, "Fernet key must be 64 url-safe base64-encoded bytes."
+      end
+
+      @signing_key    = decoded_key[0,32]
+      @encryption_key = decoded_key[32,32]
+
+      @cipher = OpenSSL::Cipher::AES256.new(:CBC)
+      @hmac   = OpenSSL::HMAC.new signing_key, OpenSSL::Digest::SHA256.new
+    end
+
+    # This is from the main bulk of the Fernet#decrypt method in the
+    # cryptography python lib:
+    #
+    #   https://github.com/pyca/cryptography/blob/ed5a7717/src/cryptography/fernet.py#L105-L122
+    #
+    # The verify bit will be described below, but the bluk of what is in the
```

Comment:
Typo `bluk`

Reply:
I don't know... I kinda like `bluk` better... has a fun ring to it!